GH-3710: Introduce commit_purge API

mpapierski opened the pull request:
Closes #3710 
This PR implements the "commit_purge" API in EE and additional machinery in the ContractRuntime component to execute it after each block.

There are also improvements to the global state to ensure that deleting an entry from the tip of the trie does not involve deserialization, and additionally, scratch trie was modified to cache the bytes of a trie rather than the trie object itself to allow lazy deserialization. This should reduce the memory footprint of a large delete operation by about 6x.

Currently, commit_purge API is called repeatedly after each block is executed with a slice of Key::EraInfo whose length is configured by chainspec's `purge_batch_size`. After several blocks, when there are no more Key::EraInfo entries, the process will be no-op, and at the next protocol version can be disabled by setting `purge_batch_size = 0`

The purge of Key::EraInfo consists of two steps:
- At the upgrade point last Key::EraInfo(activation_point_era - 1) will be copied into a Key::EraSummary.
- At the end of each executed block there will be a call to commit_purge to delete N's first Key::EraInfo entries starting from 0.

alsrdn committed:
nctl: allow override of the NCTL base URL for remote assets

The remote assets are only compiled for x86-64 which prevents running
the tests on other platforms (like ARM aarch64).
Add ability to override the base URL so that users can run the tests
using assets from a different location, or locally cached assets by
specifying a `file://<local_directory>` URL.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

Commit:
Add a regression test with a fixture.

Commit:
Fix the regression test

mpapierski committed:
Ensure EraSummary is properly written.

This also ensures no new EraInfo objects are produced, and passes all
the integration tests.

Commit:
wip refactor

Commit:
Ensure delete operation never deserializes V.

Commit:
Add commit_purge API with a failing tests

The failure comes from the ScratchTrie which caches the Trie objects,
rather than bytes. To ensure that no V is deserialized while deleting,
we're bypassing a cache, and that leads to a failure after first key is
deleted.

Commit:
Fix scratch trie to cache bytes, not tries

With a cache that operates at byte level we can still perform
optimizations such as avoiding deserialization.

Commit:
Ensure descendant iterator works

Commit:
Fix scratch trie when op didn't use get_raw.

If an operation only uses get() instead of get_raw() it may fail while
writing root to db.

Commit:
Update fixture and fix test

Fixture was regenerated on top of commit 80383403.

Commit:
Add a test to ensure no new EraInfos are written

Commit:
WIP fix

Commit:
Integrate commit purge.

Commit:
Run 11th scenario on CI

Commit:
Add batch size to chainspec

Commit:
Remove commented out code

Commit:
More cleanups

Commit:
Fix the unit test

Commit:
Do nothing if the batch_size is 0

Fraser999 commented on types/src/key.rs:
```diff
@@ -93,6 +98,7 @@ pub enum KeyTag {
     Withdraw = 8,
     Dictionary = 9,
     SystemContractRegistry = 10,
+    EraSummary = 14,
```
Shouldn't we just make this `11`, and the unpublished ones in `dev` can all get incremented?

Fraser999 commented on types/src/key.rs:
```diff
@@ -197,6 +207,9 @@ impl Display for FromStrError {
                 )
             }
+            FromStrError::EraSummary(error) => {
+                write!(f, "era-summary from string error: {}", error)
```
```suggestion
                write!(f, "era-summary-key from string error: {}", error)
```
(This will fix the failing test)

Fraser999 commented on types/src/key.rs:
```diff
@@ -57,7 +58,9 @@ pub const KEY_DICTIONARY_LENGTH: usize = 32;
 pub const DICTIONARY_ITEM_KEY_MAX_LENGTH: usize = 64;
 
-const SYSTEM_CONTRACT_REGISTRY_KEY: [u8; 32] = [0u8; 32];
+const SYSTEM_CONTRACT_REGISTRY_PADDING_KEY: [u8; 32] = [0u8; 32];
```
We started using `1111...`, `2222...` for these kinds of keys in `dev`.  Shouldn't we make this `[1u8; 32]` and update the ones in `dev` to use 2 and 3?

Fraser999 replied:
After today's discussion, can we just rename `SYSTEM_CONTRACT_REGISTRY_KEY` to something like `PADDING` and use that for both key variants?  That would be more in keeping with the intent that all key variants like this use the same padding value.

mpapierski replied:
Done

Fraser999 commented on types/src/key.rs:
```diff
@@ -316,6 +337,15 @@ impl Key {
             Err(error) => return Err(error.into()),
         }
 
+        if let Some(era_summary_padding) = input.strip_prefix(ERA_SUMMARY_PREFIX) {
+            let padded_bytes = checksummed_hex::decode(era_summary_padding)
+                .map_err(|error| FromStrError::EraSummary(error.to_string()))?;
+            let _padding: [u8; 32] = TryFrom::try_from(padded_bytes.as_ref()).map_err(|_| {
+                FromStrError::EraSummary("Failed to deserialize era summary key".to_string())
+            })?;
+            return Ok(Key::EraSummary);
+        }
+
```
I guess these should be ordered from most likely to be called to least likely.  But for now they're ordered as listed in the `Key` definition, so it'd be good to move this to the end?

mpapierski replied:
It shares common prefix (`era-summary-` with `era-`) so the longest with same prefix should be first

Fraser999 commented on types/src/key.rs:
We need to include the new variant in `impl Distribution<Key> for Standard`, and in the test `should_display_key`

mpapierski replied:
Done

Fraser999 replied:
Still not tested in `should_display_key` I think?

mpapierski replied:
Added test case

Fraser999 commented on types/src/bytesrepr.rs:
```diff
@@ -136,7 +136,15 @@ impl Display for Error {
+///
+/// Returns an error if the bytes cannot be deserialized into `T` or if not all of the input bytes
+/// are consumed in the operation.
+pub fn deserialize_from_slice<T: FromBytes>(bytes: &[u8]) -> Result<T, Error> {
```
This is already in `dev` via a different commit.  Seems like an unnecessary change to include in 1.4.14?

mpapierski replied:
I needed that in lmdb.rs to lazily construct list of descendants from a given trie bytes without making a Vec from a slice.

Fraser999 replied:
Is it really needed though?  Seems like an optimization being added to this PR which isn't required and which will cause conflicts when we merge this back to dev as the optimization already exists there.

mpapierski replied:
I prefer to leave it in, as there's no reason to create a Vec instance from slice to deserialize it. It's used in cases where I inspect a slice of bytes coming from a cache of trie bytes, then I'm deserializing real Trie based on the peeked tag. With this function, I don't need to own the borrowed data.

Fraser999 replied:
Yes, I understand the rationale.  I just think this changeset is already going to be difficult to merge to dev and I don't like that we're cherry-picking optimizations from dev when they're not required.  Let's wait for input from another reviewer?

mpapierski replied:
As agreed the code was refactored so deserialize_from_slice is not necessary in this release, and I made heap copies of the data where possible.

Fraser999 commented on types/src/system/auction/mod.rs:
```diff
@@ -554,7 +554,7 @@ pub trait Auction:
         let seigniorage_recipients = self.read_seigniorage_recipients()?;
         let base_round_reward = self.read_base_round_reward()?;
-        let era_id = detail::get_era_id(self)?;
+        // let era_id = detail::get_era_id(self)?;
```
```suggestion
```

Fraser999 commented on types/src/system/auction/providers.rs:
```diff
@@ -52,8 +52,8 @@ pub trait StorageProvider {
-    /// Records era info at the given era id.
-    fn record_era_info(&mut self, era_id: EraId, era_info: EraInfo) -> Result<(), Error>;
+    /// Records era summary.
+    fn record_era_summary(&mut self, era_summary: EraInfo) -> Result<(), Error>;
```
Technically this is breaking a public API.  We should maybe leave the method alone here (just ignoring the `era_id` in the implementation)?

On the other hand, I can't imagine this change breaking any upstream consumers since I expect the only consumers of this trait are other crates in this repo.

mpapierski replied:
Restored parameter

Fraser999 commented on node/src/types/chainspec/activation_point.rs:
```diff
@@ -29,6 +29,12 @@ pub enum ActivationPoint {
     Genesis(Timestamp),
 }
 
+impl From<EraId> for ActivationPoint {
+    fn from(v: EraId) -> Self {
+        Self::EraId(v)
+    }
+}
+
```
I don't think we'll need to keep this.  Generally we're not constructing an `ActivationPoint` outside of parsing a chainspec.toml.  The `ActivationPoint::era_id` allows us to use an `EraId` most places you're using an `ActivationPoint` and is simpler to use as we don't need to match on the activation point variants.

mpapierski replied:
Removed and changed argument from ActivationPoint to EraId

Fraser999 commented on node/src/effect/requests.rs:
```diff
@@ -716,6 +716,7 @@ pub(crate) enum ContractRuntimeRequest {
         deploys: Vec<Deploy>,
         /// The transfers for that `FinalizedBlock`
         transfers: Vec<Deploy>,
+        activation_point_block_height: Option<u64>,
```
Seems like this could be non-optional.  For genesis, it's 0, for all other protocol versions, it's the successor of the last block of the previous protocol version.

mpapierski replied:
Addressed

Fraser999 commented on node/src/effect/requests.rs:
```diff
@@ -819,6 +821,7 @@ impl Display for ContractRuntimeRequest {
                 finalized_block,
                 deploys: _,
                 transfers: _,
+                activation_point_block_height: _,
```
```suggestion
..
```

Fraser999 commented on node/src/effect/requests.rs:
```diff
@@ -807,6 +808,7 @@ pub(crate) enum ContractRuntimeRequest {
         /// The transfers for the block to execute; must correspond to the transfer and execution
         /// hashes of the `finalized_block` in that order.
         transfers: Vec<Deploy>,
+        activation_point_era_id: Option<u64>,
```
Also seems like it could be non-optional.  It should also be `activation_point_block_height` I think.

mpapierski replied:
Removed Option here and renamed

Commit:
Fix key deserialization test failure

Fraser999 commented on node/src/effect.rs:
```diff
@@ -974,6 +974,26 @@ impl<REv> EffectBuilder<REv> {
             .await
     }
 
+    pub(crate) async fn get_activation_point_era_id(self) -> Option<u64>
```
This is getting the block height, not era ID.

mpapierski replied:
Renamed to something less confusing... `get_key_block_height_for_activation_point`

Fraser999 commented on node/src/effect.rs:
```diff
@@ -974,6 +974,26 @@ impl<REv> EffectBuilder<REv> {
+    where
+        REv: From<ChainspecLoaderRequest> + From<StorageRequest>,
+    {
+        let CurrentRunInfo {
+            activation_point, ..
+        } = self.get_current_run_info().await;
+        info!(?activation_point, "get_activation_point_era_id");
+        match activation_point {
+            ActivationPoint::EraId(era_id) => self
+                .get_key_block_header_for_era_id_from_storage(era_id)
+                .await
+                .map(|block_header| block_header.height()),
+            ActivationPoint::Genesis(_timestamp) => {
+                // Nothing to do at the genesis
+                None
+            }
+        }
```
Seems like the height of the first block of the current protocol version is something which could be established in the `ChainspecLoader` and cached there, avoiding repeated calls to storage.

That way, it would be set from when the `ChainspecLoader` is finished initializing and this method doesn't need to return an `Option`.

mpapierski replied:
Seems like we do already cache this information at the request, and the indexing happens at the start of the process. Here's the retrieval under the hood https://github.com/casper-network/casper-node/blob/4c520a64b8b5bf20c19cf94b7b66b4110938ac30/node/src/components/storage.rs#L1062-L1074

mpapierski replied:
My bad, I misread this method and I was assuming it's caching the header, not the hash. I added a cache for switch block height by era id in 04de5f0a6c0c73e3994d115e9066f9086da0634a

Fraser999 commented on node/src/components/gossiper/tests.rs:
```diff
@@ -384,9 +328,76 @@ impl reactor::Reactor for Reactor {
         }
     }
 
+    fn new(
```
It would be a simpler diff if `new()` wasn't moved.

mpapierski replied:
Un-moved

Fraser999 commented on node/src/components/contract_runtime/operations.rs:
```diff
@@ -23,13 +25,52 @@ use crate::{
+    let nth_chunk: u64 = match current_height.checked_sub(activation_height) {
+        Some(nth_chunk) => nth_chunk,
+        None => {
+            // Time went backwards, programmer error, etc
```
Seems like we should at least log an error here.

mpapierski replied:
Added

Fraser999 replied:
As discussed offline (and noted in your own comment below) we agreed to change this to panicking code, e.g. `current_height.checked_sub(activation_height).expect("...")`

mpapierski replied:
My bad, I didn't check in this change. Added a `panic!` and changed `warn!` to `error!`,

Fraser999 commented on node/src/components/contract_runtime/operations.rs:
```diff
@@ -23,13 +25,52 @@ use crate::{
     storage::global_state::{CommitProvider, StateProvider},
 };
 
+fn generate_range_by_index(n: u64, m: u64, index: u64) -> Option<Range<u64>> {
+    let start = index.checked_mul(m)?;
+    let end = cmp::min(start.checked_add(m)?, n);
+    Some(start..end)
+}
```
```suggestion
fn generate_range_by_index(
    highest_era: u64,
    batch_size: u64,
    batch_index: u64,
) -> Option<Range<u64>> {
    let start = batch_index.checked_mul(batch_size)?;
    let end = cmp::min(start.checked_add(batch_size)?, highest_era);
    Some(start..end)
}
```

mpapierski replied:
Fair

Fraser999 commented on node/src/components/contract_runtime/operations.rs:
```diff
@@ -161,6 +205,63 @@ pub fn execute_finalized_block(
+                Some(keys_to_purge) => {
+                    let first_key = keys_to_purge.first().copied();
+                    let last_key = keys_to_purge.last().copied();
+                    info!(previous_block_height, %activation_point_block_height, %state_root_hash, first_key=?first_key, last_key=?last_key, "commit_purge: preparing purge config");
```
Can we help out rustfmt and ensure the log lines in this code block aren't over 100 chars pls?

mpapierski replied:
Done

Fraser999 commented on node/src/components/contract_runtime.rs:
```diff
@@ -563,10 +584,15 @@ impl ContractRuntime {
         deploys: Vec<Deploy>,
         transfers: Vec<Deploy>,
+        activation_point: ActivationPoint,
+        activation_point_era_id: Option<u64>,
```
```suggestion
        activation_point_block_height: Option<u64>,
```

mpapierski replied:
I renamed it to `key_block_height_for_activation_point` - it's not the height of the activation point, but the height of an era previous to the activation point (which I think in other places is referred to as a "key block")

Fraser999 commented on node/src/components/contract_runtime.rs:
```diff
@@ -620,9 +649,17 @@ impl ContractRuntime {
             queue.remove(&new_execution_pre_state.next_block_height)
         };
+
+        let activation_point_block_height = effect_builder.get_activation_point_era_id().await;
```
I think we only need to make this call if the `activation_point_block_height` passed into the function was `None`.

If we can change to not have optional values for these though, we don't need this call - we just copy what was passed in.

mpapierski replied:
Correct, removed unnecessary call

Fraser999 reviewed (commented):
Partial review.

Fraser999 commented on execution_engine/Cargo.toml:
```diff
@@ -49,6 +49,7 @@ uuid = { version = "0.8.1", features = ["serde", "v4"] }
 # and pwasm-utils 0.16 as upstream wasmi still depends on 0.41.0.
 # https://github.com/paritytech/wasmi/commit/f5fd480260490ff0de455017229caf7baee68195
 wasmi = "0.8.0"
+either = "1.8.1"
```
Should be in alphabetical order.

mpapierski replied:
Done

Fraser999 commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -49,6 +50,7 @@ use casper_types::{
     KeyTag, Motes, Phase, ProtocolVersion, PublicKey, RuntimeArgs, StoredValue, URef, U512,
 };
 
+use self::purge::{PurgeConfig, PurgeResult};
```
These should probably be in the `pub use self::` block below for consistency and convenience.

mpapierski replied:
Changed, and removed `pub` from `mod purge`

Fraser999 commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -442,9 +445,40 @@ where
+        // Perform global state migrations that require state.
+
+        if let Some(activation_point) = upgrade_config.activation_point() {
+            // Activation point is set to the next era after the current one.
```
```suggestion
            // The highest stored era is the immediate predecessor of the activation point.
```

mpapierski replied:
Applied

Fraser999 commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -442,9 +445,40 @@ where
+                        .force_write(Key::EraSummary, stored_value);
+                }
+                Some(_other_stored_value) => {
+                    // This should not happen as we only write EraInfo variants.
```
Shouldn't we return an error here?  Or at least log an error?

mpapierski replied:
Added an error variant in `ProtocolUpgradeError` and logged error with `error!`

Fraser999 commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -442,9 +445,40 @@ where
+                }
+                None => {
+                    // Can't find key
+                    // Most likely this chain did not yet ran an auction
```
I'd have thought this is another error case for 1.5.0, but will be the expected case for every upgrade after that?

mpapierski replied:
This could happen when you'd activate another protocol version with the same binary while the purge didn't finish yet - there will be no `EraInfo` value under `Key::EraInfo(activation_point_era_id - 1)`. I think doing nothing is a safe choice here - I just added a `warn!` message just to be able to observe this.

Fraser999 commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -2084,4 +2118,39 @@ where
+
+        match self
+            .state
+            .delete_keys(correlation_id, state_root_hash, keys_to_delete)
```
I was expecting this would need a `self.state.commit()` afterwards, but seems like each individual call to `delete` inside `delete_keys` is creating a new transaction and committing.

mpapierski replied:
Correct, it's happening internally inside the lmdb implementation. https://github.com/casper-network/casper-node/blob/582d99c9a661739fb6350c9ccc2a3aecab340658/execution_engine/src/storage/global_state/lmdb.rs#L289-L316 

Here `write_root_to_db` starts a real write transaction on a LMDB db instance

Fraser999 commented on execution_engine/src/core/engine_state/upgrade.rs:
```diff
@@ -165,7 +165,6 @@ impl From<bytesrepr::Error> for ProtocolUpgradeError {
         ProtocolUpgradeError::Bytesrepr(error)
     }
 }
-
```
Can we revert this change pls?

mpapierski replied:
Done

Fraser999 reviewed (commented):
Further comments, review still incomplete.

Commit:
Apply review comments

mpapierski commented on node/src/components/contract_runtime/operations.rs:
```diff
@@ -30,6 +32,50 @@ use casper_execution_engine::{
+        None => {
+            // Time went backwards, programmer error, etc
+            warn!(%activation_era_id, activation_height, current_height, batch_size, "unable to calculate eras to purge (activation height higher than the block height)");
+            return None;
```
Add panic/unreachable

rafal-ch commented on node/src/components/contract_runtime/operations.rs:
```diff
@@ -30,6 +32,50 @@ use casper_execution_engine::{
+/// Outcomes:
+/// * Ok(Some(range)) -- these keys should be purged
+/// * Ok(None) -- nothing to do, either done, or there is not enough eras to purge
+fn calculate_purge_eras(
```
```suggestion
fn calculate_prune_eras(
```
It seems like we prefer this process to be called "pruning", which can be seen in [other places](https://github.com/casper-network/casper-node/pull/3811/files#diff-7884e55a78e0ad8ddf73a567cf9e82cb9ff732dfa408959a7140a5444d9f3b55R470) already.

And if not, then change "prune" to "purge" in other places (for consistency).

mpapierski replied:
@Fraser999 Thoughts on renaming? There's also chainspec entry `purge_batch_size` to change which we've already communicated to test

Fraser999 replied:
I think Ed's preference was also "prune", so maybe worth changing now before the PR merges.

Fraser999 replied:
Looks like you renamed it to prune everywhere except this file! :smile:

mpapierski replied:
Of course I forgot 🤦 Fixed

rafal-ch commented on execution_engine/src/storage/global_state/in_memory.rs:
```diff
@@ -283,6 +283,29 @@ impl StateProvider for InMemoryGlobalState {
         Ok(missing_descendants)
     }
+
+    fn delete_keys(
```
Shall we consider deleting all these keys in a single transaction?

mpapierski replied:
Modified delete operation to receive a txn (just like any other operation) just like original design of delete, to be able to delete keys in a single transaction - this has also benefit for lmdb backend, so we wouldn't create multiple txs (in this case we'll reuse single Arc holding a scratch db, rather than increment it multiple times)

rafal-ch commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -505,41 +510,162 @@ where
+    let path = key_bytes;
+
+    let mut current_trie;
+    let mut current = root_bytes; //Bytes::from(root_bytes.to_bytes()?);
```
```suggestion
    let mut current = root_bytes;
```

rafal-ch commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -505,41 +510,162 @@ where
+    let mut acc: Parents<K, V> = Vec::new();
+
+    loop {
+        current_trie = if current.first() == Some(&0) {
```
Does `0` at first position have a special meaning? If so, can we used named constant instead of magic number?

mpapierski replied:
Good point, I was thinking about it. I added a `TrieTag` enum to represent it, and used it across the trie operations

rafal-ch commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -505,41 +510,162 @@ where
     // Check that tip is a leaf
     match tip {
-        Trie::Leaf { key, .. } if key == *key_to_delete => {}
+        Either::Left(_bytes)
```
```suggestion
        Either::Left(bytes)
```

Commit:
Apply @Fraser999 comments

Commit:
Apply review comments from @rafal-ch

Commit:
Add "chain_get_era_summary" RPC

Commit:
Fix incorrect GS query

Commit:
Merge remote-tracking branch 'michal/gh-3710-commit-purge' into gh-3710-commit-purge

rafal-ch commented on node/src/components/rpc_server/rpcs/chain.rs:
```diff
@@ -468,6 +475,111 @@ impl RpcWithOptionalParamsExt for GetEraInfoBySwitchBlock {
+pub struct GetEraSummary {}
+
+impl RpcWithOptionalParams for GetEraSummary {
+    const METHOD: &'static str = "chain_get_era_summary";
```
Please remember to extend out RPC testing suite in `casper-test`. Could (and should) be done in a separate PR, but please request a proper ticket.

EdHastingsCasperAssociation replied:
@RitaMAllenCA pls add ticket for this

dwerner commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -505,41 +512,167 @@ where
+                    TrieTag::from_u8(tag_u8),
+                    Some(TrieTag::Leaf),
+                    "Tip contains Leaf bytes"
+                );
```
We should return an error, as this will kill a node.

mpapierski replied:
`Either::Left` branch here implies that this tag is a `TrieTag::Leaf`, if for some reason non-leafs are inside `Either::Left` then it's a programmer's error. Here's the relevant code from `scan_raw` where we also inspect first byte: https://github.com/casper-network/casper-node/blob/6233f31fcb3eebd188b6e020f39786cd89fb2f25/execution_engine/src/storage/trie_store/operations/mod.rs#L551-L556

dwerner commented on execution_engine/src/storage/trie_store/operations/tests/delete.rs:
```diff
@@ -125,9 +145,14 @@ mod partial_tries {
-        let mut txn = environment.create_read_write_txn()?;
-        match checked_delete::<K, V, _, _, E>(correlation_id, &mut txn, store, root, key_to_delete)?
-        {
+        // let mut txn = environment.create_read_write_txn()?;
```
remove commented code.

Commit:
Apply @dwerner's review comments

Fraser999 commented on types/src/key.rs:
```diff
@@ -1025,11 +1076,12 @@ mod tests {
-            assert_eq!(*key, parsed_key);
+            assert_eq!(parsed_key, *key, "{string} (key = {key:?})");
         }
+        dbg!(Key::EraSummary.to_formatted_string());
```
Left in by mistake?

mpapierski replied:
Removed

Fraser999 commented on types/src/system/auction/providers.rs:
```diff
@@ -52,8 +52,8 @@ pub trait StorageProvider {
-    /// Records era info at the given era id.
-    fn record_era_info(&mut self, era_id: EraId, era_info: EraInfo) -> Result<(), Error>;
+    /// Records era summary.
+    fn record_era_summary(&mut self, _era_id: EraId, era_summary: EraInfo) -> Result<(), Error>;
```
As per the previous comment on this method, this is still technically a breaking change as we've renamed the method.

mpapierski replied:
Renamed back to `record_era_info`

Fraser999 commented on node/src/components/gossiper/tests.rs:
```diff
@@ -384,6 +394,7 @@ impl reactor::Reactor for Reactor {
                 self.contract_runtime
                     .handle_event(effect_builder, rng, *event),
             ),
+            Event::Chainspec(_request) => todo!(),
```
Should this be `unreachable!` rather than `todo!` or did you mean to add something more here?

mpapierski replied:
Changed to `unreachable!` - this test reactor does not use chain spec requests

Fraser999 commented on node/src/effect.rs:
```diff
@@ -974,6 +974,21 @@ impl<REv> EffectBuilder<REv> {
             .await
     }
 
+    /// Returns block height of an era previous to the activation point.
```
```suggestion
    /// Returns the key block height for the current protocol version's activation point, i.e. the
    /// height of the final block of the previous protocol version.
```

Fraser999 commented on node/src/effect.rs:
```diff
@@ -974,6 +974,21 @@ impl<REv> EffectBuilder<REv> {
+        let CurrentRunInfo {
+            activation_point, ..
+        } = self.get_current_run_info().await;
+        info!(?activation_point, "get_activation_point_era_id");
```
This should probably be removed?

mpapierski replied:
Removed

Fraser999 commented on node/src/effect.rs:
```diff
@@ -1226,6 +1241,7 @@ impl<REv> EffectBuilder<REv> {
         finalized_block: FinalizedBlock,
         deploys: Vec<Deploy>,
         transfers: Vec<Deploy>,
+        activation_point_era_id: u64,
```
```suggestion
        key_block_height_for_activation_point: u64,
```

mpapierski replied:
Changed

Fraser999 commented on node/src/effect.rs:
```diff
@@ -1256,6 +1273,7 @@ impl<REv> EffectBuilder<REv> {
         finalized_block: FinalizedBlock,
         deploys: Vec<Deploy>,
         transfers: Vec<Deploy>,
+        activation_point_block_height: u64,
```
```suggestion
        key_block_height_for_activation_point: u64,
```

mpapierski replied:
Changed

Fraser999 commented on node/src/effect/requests.rs:
```diff
@@ -716,6 +716,8 @@ pub(crate) enum ContractRuntimeRequest {
         deploys: Vec<Deploy>,
         /// The transfers for that `FinalizedBlock`
         transfers: Vec<Deploy>,
+        /// Height of the era before the activation point.
```
```suggestion
        /// The key block height for the current protocol version's activation point.
```

mpapierski replied:
Changed

Fraser999 commented on node/src/effect/requests.rs:
```diff
@@ -807,6 +809,8 @@ pub(crate) enum ContractRuntimeRequest {
         /// The transfers for the block to execute; must correspond to the transfer and execution
         /// hashes of the `finalized_block` in that order.
         transfers: Vec<Deploy>,
+        /// Height of the first block for an era previous to the activation point.
```
```suggestion
        /// The key block height for the current protocol version's activation point.
```

Commit:
Cache switch block's height by era id.

Commit:
Apply @Fraser999's review comments

Commit:
Remove dbg!

Fraser999 commented on node/src/components/rpc_server/rpcs/chain.rs:
```diff
@@ -5,7 +5,7 @@
 mod era_summary;
 
-use std::{num::ParseIntError, str};
+use std::{clone::Clone, num::ParseIntError, str};
```
I don't think we need this change.

mpapierski replied:
Removed

Commit:
Pass tx object in, rather than create it every time.

Fraser999 commented on node/src/components/rpc_server/rpcs/chain.rs:
```diff
@@ -468,6 +475,111 @@ impl RpcWithOptionalParamsExt for GetEraInfoBySwitchBlock {
+#[serde(deny_unknown_fields)]
+pub struct GetEraSummaryParams {
+    /// The block identifier.
+    pub block_identifier: BlockIdentifier,
```
Weren't we planning to use `GlobalStateIdentifier` here?  In 1.4 it only has two variants (no block height option), but in `dev` it has all three variants.

EdHastingsCasperAssociation replied:
the summary needs the block header to get the era and (not really necessary but already there) the block_hash, iirc.

Fraser999 commented on execution_engine/src/core/engine_state/purge.rs:
```diff
@@ -0,0 +1,45 @@
+    },
+}
+
+/// Represents the configuration of a protocol upgrade.
```
Copy paste error?

Fraser999 commented on execution_engine/src/core/engine_state/purge.rs:
```diff
@@ -0,0 +1,45 @@
+
+impl PurgeConfig {
+    /// Create new upgrade config.
+    #[allow(clippy::too_many_arguments)]
```
```suggestion
```

mpapierski replied:
Removed (copy & paste error)

Fraser999 commented on execution_engine/src/core/engine_state/purge.rs:
```diff
@@ -0,0 +1,45 @@
+}
+
+impl PurgeConfig {
+    /// Create new upgrade config.
```
Copy paste error?

mpapierski replied:
Ouch, yes, fixed

Fraser999 commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -442,9 +445,47 @@ where
+            // The highest stored era is the immediate predecessor of the activation point.
+            let highest_era_info_id = activation_point.saturating_sub(1);
+
+            // Start of the copy
```
I'm not sure what this comment is trying to convey?

mpapierski replied:
Removed the comment

Fraser999 commented on execution_engine/src/storage/global_state/lmdb.rs:
```diff
@@ -118,6 +119,12 @@ impl LmdbGlobalState {
     }
 }
 
+#[doc(hidden)]
+pub fn compute_empty_root_hash() -> Result<(Digest, Trie<Key, StoredValue>), error::Error> {
```
This function seems unnecessary.  I'd recommend removing it altogether, but at the very least it can be private, meaning it also doesn't need the `#[doc(hidden)]` attribute.

mpapierski replied:
Made private

Fraser999 commented on execution_engine/src/storage/trie/mod.rs:
```diff
@@ -419,35 +438,36 @@ impl<K, V> Trie<K, V> {
 /// An iterator over the descendants of a trie node.
-pub enum DescendantsIterator<'a> {
+pub enum DescendantsIterator {
     /// A leaf (zero descendants) or extension (one descendant) being iterated.
     ZeroOrOne(Option<Digest>),
     /// A pointer block being iterated.
     PointerBlock {
         /// An iterator over the non-None entries of the `PointerBlock`.
-        iter: Flatten<slice::Iter<'a, Option<Pointer>>>,
+        // iter: Flatten<slice::Iter<'a, Option<Pointer>>>,
+        iter: VecDeque<Pointer>,
```
These two changes (removing the lifetime bound and changing the type of `iter`) each constitute a breaking change to the crate's public API.

I doubt it would actually be a problem in practice if we didn't do a major version bump because of these changes, but still - not a good thing to do.  Having said that, I don't see a simple way to avoid this change.  We could refactor `write_root_to_db` to use a recursive call for example, or we could hold onto a collection of maybe parsed tries to allow their child iterators to have valid lifetimes.  Not sure either is preferable to just breaking the API though.

mpapierski replied:
Reverted this change, and refactored the code around `write_root_to_db` as agreed so the lifetime is not an issue anymore

Fraser999 commented on execution_engine/src/storage/store/mod.rs:
```diff
@@ -46,9 +48,29 @@ pub trait Store<K, V> {
+    /// This accepts a [`Cow`] object as a value to allow different implementations to choose if
+    /// they want to use owned value (i.e. put it in a cache without cloning) or the raw bytes
+    /// (write it into a persistent store).
+    fn put_raw<T>(
```
Could we have a consistent order for these methods please?  e,g get, put, get_raw, put_raw.  Same order should be followed in the impl in trie_store/lmdb.rs please.

mpapierski replied:
Reordered

Fraser999 replied:
It's not following the same order in the impl in lmbd.rs.

EdHastingsCasperAssociation replied:
I agree w/ Fraser on this...it's a nitpick but consistency is appreciated.

Fraser999 commented on execution_engine/src/storage/store/mod.rs:
```diff
@@ -46,9 +48,29 @@ pub trait Store<K, V> {
+    fn put_raw<T>(
+        &self,
+        txn: &mut T,
+        key_bytes: &[u8],
```
Can we follow the example set by `get_raw` and have this as `key: &K,` where `K: AsRef<[u8]>,`?

That way we get to have the impl for `ScratchTrieStore` use an actual `&Digest` for the key meaning we don't need the `Digest::try_from` and `debug_assert_eq!`

mpapierski replied:
Changed, good point

Fraser999 commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -528,18 +647,31 @@ where
     E: From<S::Error> + From<bytesrepr::Error>,
 {
-    let root_trie = match store.get(txn, root)? {
+    // let mut txn = environment.create_read_write_txn()?;
```
```suggestion
```

mpapierski replied:
Removed

Fraser999 commented on execution_engine_testing/test_support/src/wasm_test_builder.rs:
```diff
@@ -1293,4 +1302,43 @@ where
+
+        self.purge_results.push(result);
+        self
+    }
+    /// Returns a `Result` containing an [`PurgeResult`].
```
```suggestion
    }

    /// Returns a `Result` containing a [`PurgeResult`].
```

Fraser999 commented on execution_engine_testing/test_support/src/wasm_test_builder.rs:
```diff
@@ -1293,4 +1302,43 @@ where
+        index: usize,
+    ) -> Option<&Result<PurgeResult, engine_state::Error>> {
+        self.purge_results.get(index)
+    }
+    /// Expects a purge success.
```
```suggestion
    }

    /// Expects a purge success.
```

mpapierski replied:
Fixed

Fraser999 commented on execution_engine_testing/tests/src/lmdb_fixture.rs:
```diff
@@ -95,6 +95,12 @@ pub fn generate_fixture(
 
+    let path_to_data_lmdb = fixture_root.join("global_state").join("data.lmdb");
+    if path_to_data_lmdb.exists() {
+        eprintln!("Lmdb fixture located at {} already exists. If you need to re-generate a fixture to ensure a serialization changes are backwards compatible please make sure you are running a specific version, or a past commit. Skipping.", path_to_data_lmdb.display());
```
Can we wrap this at 100 chars pls?

Fraser999 commented on execution_engine_testing/tests/Cargo.toml:
```diff
@@ -36,7 +36,9 @@ walrus = "0.19.0"
 wat = "1.0.47"
 
 [features]
+default = []
 use-as-wasm = ["casper-engine-test-support/use-as-wasm"]
+fixture-generators = []
```
I'd rather avoid adding undocumented new features. Do we really need these changes?

mpapierski replied:
I think it's useful to keep the code that generates lmdb fixtures in the code, but I don't see a better way other than commenting the code, or removing the fixture generator

Fraser999 replied:
If it's documented how to use it then it's useful.  Without docs I don't think it's particularly helpful.

Also, couldn't we use an env var for this to save cluttering the manifest and to avoid recompiling?

mpapierski replied:
Good point, removed the feature flag and added `RUN_FIXTURE_GENERATORS` env var (although it's unfortunate you can't skip/ignore tests based on env var)

Commit:
Calculate height for previous era.

This regression was introduced 04de5f0a and resulted in purge starting
from next switch block after activation, instead of starting within
first block activation.

Fraser999 commented on types/src/key.rs:
```diff
@@ -1109,21 +1170,31 @@ mod tests {
-                base16::encode_lower(&SYSTEM_CONTRACT_REGISTRY_KEY)
-            ),
+        let expected_json = &[
+            json!({ "Account": format!("account-hash-{}", HEX_STRING) }),
+            json!({ "Hash": format!("hash-{}", HEX_STRING) }),
+            json!({ "URef": format!("uref-{}-001", HEX_STRING) }),
+            json!({ "Transfer": format!("transfer-{}", HEX_STRING) }),
+            json!({ "DeployInfo": format!("deploy-{}", HEX_STRING) }),
+            json!({"EraInfo": "era-42" }),
+            json!({ "Balance": format!("balance-{}", HEX_STRING) }),
+            json!({ "Bid": format!("bid-{}", HEX_STRING) }),
+            json!({ "Withdraw": format!("withdraw-{}", HEX_STRING) }),
+            json!({ "Dictionary": format!("dictionary-{}", HEX_STRING) }),
+            json!({
+                "SystemContractRegistry":
+                    format!(
+                        "system-contract-registry-{}",
+                        base16::encode_lower(&SYSTEM_CONTRACT_REGISTRY_PADDING_KEY)
+                    )
+            }),
+            json!({
+                "EraSummary":
+                    format!(
+                        "era-summary-{}",
+                        base16::encode_lower(&ERA_SUMMARY_PADDING_KEY)
+                    )
+            }),
```
Why was this change required?  I prefer the previous approach as it's more obvious what the JSON-encoded form should be.

mpapierski replied:
I added a new assertion with JSON-encoded string, and then I changed this test to compare JSON values, as I had some random unexpected issues when writing JSON by hand. I like the type safety and the fact that I don't need to worry about specifics of a JSON serializer when doing such comparisons (would it sort the keys when encoding? etc)

rafal-ch replied:
I lean towards accepting the change. Even though json strings in this test case are simple, I prefer the comparison that prevents false-negatives when, for example, the serialization mechanism changes and we get different whitespace characters in the output, etc.
The test should not fail as long as the logic structure of the two jsons match.

Fraser999 commented on node/src/components/storage.rs:
```diff
@@ -922,6 +927,11 @@ impl Storage {
             } => responder
                 .respond(self.store_finalized_approvals(deploy_hash, finalized_approvals)?)
                 .ignore(),
+            StorageRequest::GetSwitchBlockHeightAtEraId { era_id, responder } => responder
+                .respond(
+                    self.get_switch_block_height_by_era_id(&mut self.env.begin_ro_txn()?, era_id)?,
+                )
+                .ignore(),
```
We can avoid the `RefCell` if we just inline the code something like this:
```suggestion
            StorageRequest::GetSwitchBlockHeightAtEraId { era_id, responder } => {
                let mut txn = self.env.begin_ro_txn()?;
                if let Some(block_height) = self.switch_block_height_by_era_id_cache.get(&era_id) {
                    return Ok(responder.respond(Some(*block_height)).ignore());
                }
                let block_header = match self.get_switch_block_header_by_era_id(&mut txn, era_id)? {
                    Some(block_header) => block_header,
                    None => return Ok(responder.respond(None).ignore()),
                };
                self.switch_block_height_by_era_id_cache
                    .insert(era_id, block_header.height());
                responder.respond(Some(block_header.height())).ignore()
            }
```

mpapierski replied:
Nice one, added. I didn't see how I can remove RefCell at first

Fraser999 commented on node/src/components/storage.rs:
```diff
@@ -370,6 +371,9 @@ pub struct Storage {
     enable_mem_deduplication: bool,
     /// Pool of loaded items.
     deploy_cache: BlobCache<<Deploy as Item>::Id>,
+    /// A map of era ID to a block height.
```
```suggestion
    /// A map of era ID to the height of the switch block for that era.
```

mpapierski replied:
Applied

Fraser999 commented on node/src/components/storage.rs:
```diff
@@ -370,6 +371,9 @@ pub struct Storage {
     /// Pool of loaded items.
     deploy_cache: BlobCache<<Deploy as Item>::Id>,
+    /// A map of era ID to a block height.
+    #[data_size(skip)]
```
We can remove this if we get rid of the `RefCell`.

mpapierski replied:
Done

Fraser999 commented on node/src/components/rpc_server/rpcs/docs.rs:
```diff
@@ -100,6 +100,7 @@ pub(crate) static OPEN_RPC_SCHEMA: Lazy<OpenRpcSchema> = Lazy::new(|| {
     schema.push_with_optional_params::<GetAuctionInfo>(
         "returns the bids and validators as of either a specific block (by height or hash), or the most recently added block",
     );
+    schema.push_with_optional_params::<GetEraSummary>("returns the era summary at either a specific block (by height or hash), or the most recently added block");
```
We'll need to update the description here if we change to use `GlobalStateIdentifier`.

mpapierski replied:
Will deal with this as agreed offline

mpapierski replied:
As discussed, we will stick with BlockIdentifier because we can't go from `GlobalStateIdentifier::StateRootHash` to a get a `Block` to fill `EraSummary.block_hash`  as state root hashes aren't 1:1 with blocks - if a block does not execute anything then it will carry forward previous state root hash. @darthsiroftardis and @EdHastingsCasperLabs agrees on that

Fraser999 commented on node/src/components/linear_chain_sync/blocks.rs:
```diff
@@ -152,6 +153,10 @@ pub(super) async fn execute_block<REv>(
         }
     }
 
+    let activation_point_block_height = effect_builder
```
```suggestion
    let key_block_height_for_activation_point = effect_builder
```

mpapierski replied:
Applied

Fraser999 commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -505,15 +511,128 @@ where
+}
+
+/// Just like scan, however we don't parse the tip.
+fn scan_raw<K, V, T, S, E>(
```
This is largely a copy-paste of `scan` - I was wondering if we could either get rid of `scan`, or implement it like
```rust
    let raw = scan_raw::<K, V, T, S, E>(_correlation_id, txn, store, key_bytes, root_bytes)?;
    match raw.tip {
        Either::Left(bytes) => {
            let tip = bytesrepr::deserialize(bytes.to_vec())?;
            Ok(TrieScan::new(tip, raw.parents))
        }
        Either::Right(tip) => Ok(TrieScan::new(tip, raw.parents)),
    }
```

mpapierski replied:
Good point, changed

Fraser999 commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -505,15 +511,128 @@ where
+
+/// Just like scan, however we don't parse the tip.
+fn scan_raw<K, V, T, S, E>(
+    _correlation_id: CorrelationId,
```
We don't need this parameter do we?

mpapierski replied:
I guess not, I don't remember if it was ever used. I had to modify `write` and `delete` and rename parameter to `_correlation_id` as it's unused.

Fraser999 commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -505,15 +511,128 @@ where
+        current_trie = if current_trie_tag == Some(TrieTag::Leaf) {
+            return Ok(TrieScanRaw::new(Either::Left(current), acc));
+        } else {
+            let (deserialized, _) = Trie::<K, V>::from_bytes(&current)?;
```
Safer to use `bytesrepr::deserialize` so we are sure there are no unused bytes left over.

mpapierski replied:
Done already

Fraser999 commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -528,18 +647,31 @@ where
+            if {
+                // Partially deserialize a key of a leaf node to ensure that we can only continue if
+                // the key matches what we're looking for.
+                let ((tag_u8, key), _rem): ((u8, K), _) = FromBytes::from_bytes(&bytes)?;
```
I would be good to have a consistent way to do this tag inspection.  For example, in other places we have
```rust
let trie_tag = trie_bytes.first().and_then(|byte| TrieTag::from_u8(*byte));
```

mpapierski replied:
I could probably put this inside a function (`fn lazy_trie_deserialize_leaf_key(bytes) -> Result<Option<K>>` or similar) and delete this outer match statement completely, but I think it's good enough and relies on the fact that `Either::Left` has leaf bytes already, what do you think?

mpapierski replied:
Also this:

```rust
let trie_tag = trie_bytes.first().and_then(|byte| TrieTag::from_u8(*byte));
```

I've put it in a separate function

Fraser999 commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -528,18 +647,31 @@ where
+                assert_eq!(
+                    TrieTag::from_u8(tag_u8),
+                    Some(TrieTag::Leaf),
+                    "Tip contains Leaf bytes"
```
```suggestion
                    "Tip should contain leaf bytes, but has tag {:?}",
                    trie_tag
```

mpapierski replied:
Applied

Fraser999 commented on ci/nctl_upgrade.sh:
We should add this test to the nightlies [here](https://github.com/casper-network/casper-node/blob/35e08b51b043c2252f670cbb4dc8eb0a6f06a11d/ci/nightly-test.sh#L67-L77).

mpapierski replied:
Added

Commit:
Addressing big chunk of @Fraser999's comments

Fraser999 commented on utils/nctl/sh/scenarios-upgrades/upgrade_scenario_11.sh:
```diff
@@ -0,0 +1,204 @@
+    local CURRENT_ERA_ID=$(jq -r '.header.era_id' <<< "$BLOCK")
+    local SEQ_END=$(($CURRENT_ERA_ID-1))
+    local STATE_ROOT_HASH=$(jq -r '.header.state_root_hash' <<< "$BLOCK")
+    echo "state root hash $STATE_ROOT_HASH"
```
Almost all `echo` calls here should be `log`s.

mpapierski replied:
Applied

Fraser999 commented on utils/nctl/sh/scenarios-upgrades/upgrade_scenario_11.sh:
```diff
@@ -0,0 +1,204 @@
+# Synopsis.
+# ----------------------------------------------------------------
+
+# 1. Start v1 running at ProtocolVersion 1_3_0 commit.
```
I don't think it runs at that version initially?

mpapierski replied:
I'm not sure about it - I think it's 1.4.5? I updated the comment

Fraser999 commented on utils/nctl/sh/scenarios-upgrades/upgrade_scenario_11.sh:
```diff
@@ -0,0 +1,204 @@
+
+    log "... awaiting era"
+
+    # previous protocol had era id of 1, and new verison apparently activates at era 3
```
```suggestion
    # previous protocol had era id of 1, and new version apparently activates at era 3
```

mpapierski replied:
Fixed

Fraser999 commented on utils/nctl/sh/scenarios-upgrades/upgrade_scenario_11.sh:
```diff
@@ -0,0 +1,204 @@
+    await_n_eras '2' 'true' '5.0'
+    log "... awaiting multiple blocks"
+    await_n_blocks 5
+
+
```
```suggestion
```

mpapierski replied:
Fixed

Fraser999 commented on utils/nctl/sh/scenarios-upgrades/upgrade_scenario_11.sh:
```diff
@@ -0,0 +1,204 @@
+}
+
+function _step_05() {
+    log "... asserting"
```
Did you mean to add more info here?

mpapierski replied:
Updated with more info

Fraser999 commented on utils/nctl/sh/scenarios-upgrades/upgrade_scenario_11.sh:
```diff
@@ -0,0 +1,204 @@
+    local SEQ=$(($CURRENT_ERA_ID-1))
+    local STATE_ROOT_HASH=$(jq -r '.header.state_root_hash' <<< "$BLOCK")
+    echo "state root hash $STATE_ROOT_HASH"
+    local total=0
```
```suggestion
    local TOTAL=0
```

mpapierski replied:
Fixed

Fraser999 commented on utils/nctl/sh/scenarios-upgrades/upgrade_scenario_11.sh:
```diff
@@ -0,0 +1,204 @@
+        fi
+    done
+    set -e
+    echo Total era infos found $total for a total of $CURRENT_ERA_ID era
```
```suggestion
    log "found $TOTAL era infos for a total of $CURRENT_ERA_ID eras"
```

mpapierski replied:
Updated

Fraser999 commented on node/src/types/chainspec/core_config.rs:
```diff
@@ -36,6 +36,8 @@ pub struct CoreConfig {
     pub(crate) max_stored_value_size: u32,
     /// The minimum bound of motes that can be delegated to a validator.
     pub(crate) minimum_delegation_amount: u64,
+    /// Global state purge batch size (0 means the feature is off in the current protocol version).
```
```suggestion
    /// Global state prune batch size (0 means the feature is off in the current protocol version).
```

mpapierski replied:
Fixed

Fraser999 commented on execution_engine/src/storage/trie_store/operations/mod.rs:
```diff
@@ -505,15 +512,127 @@ where
-#[allow(unused)]
-fn delete<K, V, T, S, E>(
+/// Delete provided key from a global state so it is not reachable from a resulting state root hash.
+pub fn delete<K, V, T, S, E>(
```
```suggestion
pub(crate) fn delete<K, V, T, S, E>(
```

mpapierski replied:
Did this - it's unfortunate DeleteResult was pub all the time though but we'll deal with it in #3828

Fraser999 commented on execution_engine/src/storage/trie_store/lmdb.rs:
```diff
@@ -199,34 +203,32 @@ impl ScratchTrieStore {
     /// Writes only tries which are both under the given `state_root` and dirty to the underlying db
     /// while maintaining the invariant that children must be written before parent nodes.
```
```suggestion
```

mpapierski replied:
Applied

Fraser999 commented on execution_engine/src/core/engine_state/prune.rs:
```diff
@@ -0,0 +1,44 @@
+use casper_hashing::Digest;
+use casper_types::Key;
+
+/// Represents a successfully performed prune.
```
```suggestion
/// The result of performing a prune.
```

mpapierski replied:
Did this (just to let you know `UpgradeResult` etc they have the same `Represents a successfully performed XYZ` comments)

Fraser999 replied:
Maybe worth updating their comments too then if they say they represent success, but contain variants which look like failure cases?  (Not in this PR - I just mean some time in the future).

Commit:
Apply remaining @Fraser999's comments

Fraser999 commented on utils/nctl/sh/scenarios-upgrades/upgrade_scenario_11.sh:
```diff
@@ -0,0 +1,202 @@
+
+for ARGUMENT in "$@"
+do
+    KEY=$(log "$ARGUMENT" | cut -f1 -d=)
+    VALUE=$(log "$ARGUMENT" | cut -f2 -d=)
```
```suggestion
    KEY=$(echo "$ARGUMENT" | cut -f1 -d=)
    VALUE=$(echo "$ARGUMENT" | cut -f2 -d=)
```

mpapierski replied:
Fixed

Fraser999 reviewed (approved):
Optimistic approval, given the one final comment on upgrade_scenario_11.sh is addressed.

rafal-ch commented:
I'll be doing a second review today, but nevertheless, we need a stamp from @dwerner and/or @EdHastingsCasperLabs on this.

Commit:
Apply last remaining items from @Fraser999

EdHastingsCasperAssociation commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -442,9 +445,45 @@ where
+                    ));
+                }
+                None => {
+                    // Can't find key
+                    // Most likely this chain did not yet ran an auction, or recently completed a
+                    // prune
```
I don't think this comment is accurate, and while it is basically harmless to do so I don't think it is necessary to log this circumstance as a `warn`. 

As far as I recall, a new network w/ genesis at this binary or higher would never have such an entry (correct behavior). An existing network upgraded into this binary _must_ have such an entry. In a future binary we will either remove this handling altogether, or if it is left in place it would essentially translate to noop on networks that had already previously gone thru a migration. 

Am i forgetting something?

mpapierski replied:
Your comment is correct, the `warn!` here was useful for testing early in the development to get the math right. I removed the warning.

EdHastingsCasperAssociation commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -2084,4 +2123,39 @@ where
     }
+
+    /// Commit a prune of leaf nodes from the tip of the merkle trie.
+    pub fn commit_prune(
```
nitpick, but could you put this new func after commit_genesis and commit_upgrade for purposes of logical grouping?

mpapierski replied:
Done

EdHastingsCasperAssociation commented on execution_engine/src/core/runtime/externals.rs:
```diff
@@ -969,9 +969,9 @@ where
                 scoped_instrumenter.add_property("era_info_size", era_info_size.to_string());
-                let era_id: EraId = self.t_from_mem(era_id_ptr, era_id_size)?;
+                let _era_id: EraId = self.t_from_mem(era_id_ptr, era_id_size)?;
                 let era_info: EraInfo = self.t_from_mem(era_info_ptr, era_info_size)?;
-                self.record_era_info(era_id, era_info)?;
+                self.record_era_summary(era_info)?;
```
good job

EdHastingsCasperAssociation commented on execution_engine/src/core/runtime_context/mod.rs:
```diff
@@ -319,6 +319,10 @@ where
                 Err(Error::RemoveKeyFailure(RemoveKeyFailure::PermissionDenied))
             }
+            Key::EraSummary => {
+                self.named_keys.remove(name);
+                Ok(())
```
at some point we ought to | together the various self.named_keys.remove(name); Ok(()) arms.

EdHastingsCasperAssociation commented on execution_engine/src/core/runtime_context/mod.rs:
```diff
@@ -566,8 +570,8 @@ where
     /// Write an era info instance to the global state.
-    pub fn write_era_info(&mut self, key: Key, value: EraInfo) {
-        if let Key::EraInfo(_) = key {
+    pub fn write_era_summary(&mut self, key: Key, value: EraInfo) {
+        if let Key::EraSummary = key {
```
i don't think we can get away w/ changing the name of this public function in 1.x

instead, consider making a new function with the new behavior and forward the old function to it, mark the old fn as deprecated, and todo! remove it in 2.0

mpapierski replied:
Good call, I renamed it back (although I think we would be fine, as we don't reimport RuntimeContext as pub)

EdHastingsCasperAssociation commented on execution_engine/src/core/runtime_context/mod.rs:
```diff
@@ -767,6 +771,7 @@ where
                 false
             }
             Key::SystemContractRegistry => false,
+            Key::EraSummary => true,
```
is there a cltyped for it?

EdHastingsCasperAssociation replied:
ok, i see it...era_info already had it:
```
impl CLTyped for EraInfo {
    fn cl_type() -> CLType {
        CLType::List(Box::new(SeigniorageAllocation::cl_type()))
    }
}
```

( SeigniorageAllocation::cl_type() == Any )

EdHastingsCasperAssociation commented on execution_engine/src/storage/trie/mod.rs:
```diff
@@ -335,6 +338,24 @@ impl ::std::fmt::Debug for PointerBlock {
+/// Represents all possible serialization tags for a [`Trie`] enum.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, FromPrimitive, ToPrimitive)]
+#[repr(u8)]
+pub(crate) enum TrieTag {
+    /// Represents a tag for a [`Trie::Leaf`] variant.
+    Leaf = 0,
+    /// Represents a tag for a [`Trie::Node`] variant.
+    Node = 1,
+    /// Represents a tag for a [`Trie::Extension`] variant.
+    Extension = 2,
+}
```
good job

EdHastingsCasperAssociation commented on node/src/components/contract_runtime/operations.rs:
```diff
@@ -161,6 +208,70 @@ pub fn execute_finalized_block(
+                let prune_config = PruneConfig::new(state_root_hash, keys_to_prune);
+                match engine_state.commit_prune(CorrelationId::new(), prune_config) {
+                    Ok(PruneResult::RootNotFound) => {
+                        error!(
+                            previous_block_height,
+                            %state_root_hash,
+                            "commit_prune: root not found");
```
shouldn't we kill the node in this case?

mpapierski replied:
I can do this and do a panic. It's impossible to hit this case. CC @Fraser999

EdHastingsCasperAssociation commented on types/src/key.rs:
```diff
@@ -585,7 +620,10 @@ impl ToBytes for Key {
                 result.append(&mut addr.to_bytes()?);
             }
             Key::SystemContractRegistry => {
-                result.append(&mut SYSTEM_CONTRACT_REGISTRY_KEY.to_bytes()?)
+                result.append(&mut PADDING_BYTES.to_bytes()?)
+            }
+            Key::EraSummary => {
+                result.append(&mut PADDING_BYTES.to_bytes()?);
             }
```
maybe just | these arms together?

mpapierski replied:
Did this

EdHastingsCasperAssociation reviewed (approved):
What a monster. See nitpicks but LGTM, congrats to all involved. Lets test the heck out it.

Commit:
Merge branch 'release-1.4.14' into gh-3710-commit-purge

Commit:
Apply @EdHastingsCasperLabs feedback.

mpapierski commented:
bors r+
